applications/maplab-server-node/src/maplab-server-node.cc

```diff
@@ -978,4 +1024,233 @@ void MaplabServerNode::publishDenseMap() {
+  }
+  CHECK_EQ(num_matching_missions, 1u);
+  CHECK(mission_to_delete.isValid());
+
+  // Blacklist the mission, this will delete it at the end of the merging
+  // iteration and prevent subsequent submap updates of this mission from
+  // being merged.
+  {
+    std::lock_guard<std::mutex> lock(blacklisted_missions_mutex_);
+    blacklisted_missions_[mission_to_delete] = robot_name_of_mission;
+
+    ss << "Will delete and blacklist mission " << mission_to_delete.hexString();
+    *status_message = ss.str();
+    LOG(INFO) << "[MaplabServerNode] " << *status_message;
+    return true;
+  }
+}
+
+bool MaplabServerNode::deleteAllRobotMissions(
+    const std::string& robot_name, std::string* status_message) {
+  CHECK_NOTNULL(status_message);
+
+  std::stringstream ss;
+
+  if (robot_name.empty()) {
+    ss << "Robot name is empty, cannot find associated missions to delete "
+       << "them!";
+    *status_message = ss.str();
+    LOG(ERROR) << "[MaplabServerNode] " << *status_message;
+    return false;
+  }
+
+  // Retrieve all missions associated with this robot.
+  uint32_t num_matching_missions = 0u;
+  std::unordered_set<vi_map::MissionId> missions_to_delete;
+  {
+    std::lock_guard<std::mutex> lock(robot_to_mission_id_map_mutex_);
+    for (const auto& kv : mission_id_to_robot_map_) {
+      if (kv.second == robot_name) {
+        CHECK(kv.first.isValid());
+        missions_to_delete.insert(kv.first);
+        ++num_matching_missions;
+      }
+    }
+  }
+
+  if (num_matching_missions == 0u) {
+    ss << "No mission matches the provided robot name "
+       << "('" << robot_name << "')";
+    *status_message = ss.str();
+    LOG(ERROR) << "[MaplabServerNode] " << *status_message;
+    return false;
+  }
+
+  // Blacklist the mission, this will delete it at the end of the merging
+  // iteration and prevent subsequent submap updates of this mission from
+  // being merged.
+  {
+    std::lock_guard<std::mutex> lock(blacklisted_missions_mutex_);
+    ss << "Will delete and blacklist all missions of robot " << robot_name
+       << ": ";
+    for (const vi_map::MissionId& mission_to_delete : missions_to_delete) {
+      *status_message += mission_to_delete.hexString() + " ";
+      blacklisted_missions_[mission_to_delete] = robot_name;
+    }
+    *status_message = ss.str();
+    LOG(INFO) << "[MaplabServerNode] " << *status_message;
+    return true;
+  }
+}
+
+bool MaplabServerNode::deleteBlacklistedMissions() {
+  if (!map_manager_.hasMap(kMergedMapKey)) {
+    return false;
+  }
+  uint32_t num_missions_in_merged_map_after_deletion;
+  {
+    vi_map::VIMapManager::MapWriteAccess merged_map =
+        map_manager_.getMapWriteAccess(kMergedMapKey);
+
+    // Make copy of blacklisted missions to avoid repeatedly locking a
+    // potentially changing blacklist.
+    std::unordered_map<vi_map::MissionId, std::string>
+        blacklisted_missions_copy;
+    {
+      std::lock_guard<std::mutex> lock(blacklisted_missions_mutex_);
```

Comment:
is there a reason to keep all the blacklisted missions in the structure or do you delete them somewhere? otherwise, we would try to delete the already deleted missions which might be dangerous with the `std::find`, `erase` below.